src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts

```diff
@@ -0,0 +1,243 @@
+  const spec = {} as any;
+
+  const legendPosition = visParams.legendPosition;
+
+  spec.$schema = 'https://vega.github.io/schema/vega-lite/v5.json';
+  spec.data = {
+    values: datatable.rows,
+  };
+  spec.config = {
+    view: {
+      stroke: null,
+    },
+    concat: {
+      spacing: 0,
+    },
+    // the circle timeline representing annotations
+    circle: {
+      color: 'blue',
+    },
+    // the vertical line when user hovers over an annotation circle
+    rule: {
+      color: 'red',
+    },
+    legend: {
+      orient: legendPosition,
+    },
+  };
+
+  // Get the valueAxes data and generate a map to easily fetch the different valueAxes data
+  const valueAxis = {};
+  visParams?.valueAxes.forEach((yAxis: { id: { toString: () => string | number } }) => {
+    // @ts-ignore
+    valueAxis[yAxis.id.toString()] = yAxis;
+  });
+
+  spec.layer = [] as any[];
+
+  if (datatable.rows.length > 0 && dimensions.x != null) {
+    const xAxisId = getXAxisId(dimensions, datatable.columns);
+    const xAxisTitle = cleanString(dimensions.x.label);
+    const startTime = new Date(dimensions.x.params.bounds.min).valueOf();
+    const endTime = new Date(dimensions.x.params.bounds.max).valueOf();
+    let skip = 0;
+    datatable.columns.forEach((column, index) => {
+      // Check if its not xAxis column data
+      if (column.meta?.aggConfigParams?.interval != null) {
+        skip++;
+      } else {
+        const currentSeriesParams = visParams.seriesParams[index - skip];
+        const currentValueAxis =
+          // @ts-ignore
+          valueAxis[currentSeriesParams.valueAxis.toString()];
+        let tooltip: Array<{ field: string; type: string; title: string }> = [];
+        if (visParams.addTooltip) {
+          tooltip = [
+            { field: xAxisId, type: 'temporal', title: xAxisTitle },
+            { field: column.id, type: 'quantitative', title: column.name },
+          ];
+        }
+        spec.layer.push({
+          mark: {
+            // Possible types are: line, area, histogram. The eligibility checker will
+            // prevent area and histogram (though area works in vega-lite)
+            type: currentSeriesParams.type,
+            // Possible types: linear, cardinal, step-after. All of these types work in vega-lite
+            interpolate: currentSeriesParams.interpolate,
+            // The possible values is any number, which matches what vega-lite supports
+            strokeWidth: currentSeriesParams.lineWidth,
+            // this corresponds to showing the dots in the visbuilder for each data point
+            point: currentSeriesParams.showCircles,
+          },
+          encoding: {
+            x: {
+              axis: {
+                title: xAxisTitle,
+                grid: visParams.grid.categoryLines,
+              },
+              field: xAxisId,
+              type: 'temporal',
+              scale: {
+                domain: [startTime, endTime],
```

Comment:
just to confirm - this won't shrink the chart bounds to fit such that there is no empty positions at the beginning and end of the chart, right? I see you had a comment referencing a different approach here https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3106#discussion_r1103699002

Also, I think we still need time buckets even if there is empty data from the charts (empty rows). This is so the events can still fit into such empty buckets if that is the closest bucket based on their timestamps. Typically this wouldn't be the case with associated plugin resources, but because we are allowing unrelated resources to be added to the charts, it's possible that we have events even with empty data shown for a particular vis

Reply:
If you weren't able to find a good solution via agg params, I think we will need a helper fn and artificially add rows to fit the full bounds.

Reply:
Discussed offline - this won't shrink the chart bounds. We can keep the domain as temporal.

As a follow up I'll need to create a helper fn to add empty rows for potential event datapoints

Reply:
I thought that the previous approach was needed since there were some local issues with using scale and domain. The line charts actually do something very similar and dont create extra buckets. We will need to create a helper fn to add rows as changing agg params can cause a lot of unknown side effects.

Reply:
Yes, just adding rows after-the-fact is what we need, we should not touch the agg params